### PR TITLE
WIP: merge/foundation release/1.10.2 resolved

### PR DESCRIPTION
**Pull request by meowsbits**
Merges foundation at `c2d2f4ed8f`.

See [related Tracking Issue](https://github.com/etclabscore/core-geth/issues/365) for TODOs and review status.

### CODE REVIEW

**Comment by meowsbits** [comment]
Running https://github.com/etclabscore/core-geth/pull/348/commits/9e217469222875fac2ed6a34ef693fd1df2dd83a on ETC.

```
/home/ia/go/src/github.com/ethereum/go-ethereum/.builds/systemd/classic/geth \
        --classic \
        --syncmode=fast \
        --http \
        --http.port 8061 \
        --http.api=debug,eth,web3,net \
        --http.corsdomain=\'*\' \
        --port 30361 \
        --debug --verbosity 4 \
        --datadir /home/ia/SAMSUNG_T5/e/clients/core-geth/classic \
        --ethash.dagdir /home/ia/SAMSUNG_T5/e/clients/core-geth/classic/.ethash \
        --ipcpath /tmp/core-geth_classic.ipc \
```

```
Mar 24 07:24:37 ubp52 sh[22547]: INFO [03-24|07:24:37.729|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=45 static=0
Mar 24 07:24:47 ubp52 sh[22547]: INFO [03-24|07:24:47.777|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=54 static=0
Mar 24 07:24:57 ubp52 sh[22547]: INFO [03-24|07:24:57.904|p2p/dial.go:339]                    Looking for peers                        peercount=1 tried=58 static=0
Mar 24 07:25:08 ubp52 sh[22547]: INFO [03-24|07:25:08.022|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=61 static=0
Mar 24 07:25:18 ubp52 sh[22547]: INFO [03-24|07:25:18.024|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=50 static=0
Mar 24 07:25:21 ubp52 sh[22547]: INFO [03-24|07:25:21.529|eth/downloader/downloader.go:382]   Block synchronisation started
Mar 24 07:25:27 ubp52 sh[22547]: INFO [03-24|07:25:27.901|eth/downloader/queue.go:391]        Downloader queue stats                   receiptTasks=0 blockTasks=361 itemSize=288.77B throttle=8192
Mar 24 07:25:28 ubp52 sh[22547]: INFO [03-24|07:25:28.101|p2p/dial.go:339]                    Looking for peers                        peercount=1 tried=65 static=0
Mar 24 07:25:30 ubp52 sh[22547]: WARN [03-24|07:25:30.763|core/rawdb/accessors_chain.go:764]  Failed to load old bad blocks            error="leveldb: not found"
Mar 24 07:25:30 ubp52 sh[22547]: ERROR[03-24|07:25:30.763|core/blockchain.go:2609]
Mar 24 07:25:30 ubp52 sh[22547]: ########## BAD BLOCK #########
Mar 24 07:25:30 ubp52 sh[22547]: Chain config: NetworkID: 1, ChainID: 61 Engine: ethash ECBP1100: 11380000 EIP1014: 9573000 EIP1052: 9573000 EIP1108: 10500839 EIP1344: 10500839 EIP140: 8772000 EIP145: 9573000 EIP150: 2500000 EIP152: 10500839 EIP155: 3000000 EIP160: 3000000 EIP161abc: 8772000 EIP161d: 8772000 EIP170: 8772000 EIP1884: 10500839 EIP198: 8772000 EIP2028: 10500839 EIP211: 8772000 EIP212: 8772000 EIP213: 8772000 EIP214: 8772000 EIP2200: 10500839 EIP2: 1150000 EIP658: 8772000 EIP7: 1150000 EthashECIP1010Continue: 5000000 EthashECIP1010Pause: 3000000 EthashECIP1017: 5000000 EthashECIP1041: 5900000 EthashECIP1099: 11700000 EthashEIP100B: 8772000 EthashHomestead: 1150000
Mar 24 07:25:30 ubp52 sh[22547]: Number: 12345334
Mar 24 07:25:30 ubp52 sh[22547]: Hash: 0xceeea5817fb17b310be4fcf3d849df023d4247729be9ef63e5e4f92008edc0de
Mar 24 07:25:30 ubp52 sh[22547]: Error: invalid mix digest
Mar 24 07:25:30 ubp52 sh[22547]: ##############################
Mar 24 07:25:30 ubp52 sh[22547]:  
Mar 24 07:25:30 ubp52 sh[22547]: WARN [03-24|07:25:30.763|eth/downloader/downloader.go:352]   Synchronisation failed, dropping peer    peer=2a62660cce76343cf2cbb563a6b9032fddae1158a9efd83a5d0a98365e1fac0d err="retrieved hash chain is invalid: invalid mix digest"
Mar 24 07:25:30 ubp52 sh[22547]: ERROR[03-24|07:25:30.764|eth/handler.go:368]                 Ethereum peer removal failed             peer=2a62660c err="peer not registered"
Mar 24 07:25:38 ubp52 sh[22547]: INFO [03-24|07:25:38.187|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=55 static=0
Mar 24 07:25:48 ubp52 sh[22547]: INFO [03-24|07:25:48.320|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=199 static=0
Mar 24 07:25:58 ubp52 sh[22547]: INFO [03-24|07:25:58.354|p2p/dial.go:339]                    Looking for peers                        peercount=0 tried=182 static=0
Mar 24 07:26:08 ubp52 sh[22547]: INFO [03-24|07:26:08.378|p2p/dial.go:339]                    Looking for peers                        peercount=2 tried=74  static=0
```

Tried rolling back to 12_000_000, but no joy:

```
Mar 24 07:40:51 ubp52 sh[24111]: WARN [03-24|07:40:51.216|eth/downloader/downloader.go:1659]  Invalid header encountered               number=12000024 hash="e43eda…ad483a" parent="4d6e10…42a99c" err="invalid mix digest"
Mar 24 07:40:51 ubp52 sh[24111]: WARN [03-24|07:40:51.216|core/blockchain.go:623]             Rewinding blockchain                     target=12000000
Mar 24 07:40:51 ubp52 sh[24111]: INFO [03-24|07:40:51.216|core/blockchain.go:477]             Loaded most recent local header          number=12000000 hash="4df1f3…2702fb" td=1070480238291937640602 age=2mo1w4h
Mar 24 07:40:51 ubp52 sh[24111]: INFO [03-24|07:40:51.216|core/blockchain.go:478]             Loaded most recent local full block      number=0        hash="d4e567…cb8fa3" td=17179869184            age=51y11mo2w
Mar 24 07:40:51 ubp52 sh[24111]: INFO [03-24|07:40:51.216|core/blockchain.go:479]             Loaded most recent local fast block      number=12000000 hash="4df1f3…2702fb" td=1070480238291937640602 age=2mo1w4h
Mar 24 07:40:51 ubp52 sh[24111]: INFO [03-24|07:40:51.216|core/blockchain.go:481]             Loaded last fast-sync pivot marker       number=12419347
Mar 24 07:40:51 ubp52 sh[24111]: WARN [03-24|07:40:51.216|eth/downloader/downloader.go:1560]  Rolled back chain segment                header=12000000->12000000 fast=12000000->12000000 block=0->0 reason="invalid mix digest"
Mar 24 07:41:32 ubp52 sh[24111]: WARN [03-24|07:41:32.558|eth/downloader/downloader.go:352]   Synchronisation failed, dropping peer    peer=818e05481c883d1bfe49aa680a587226459f4c80a7ab44086ce7c1a4dfd68d15 err="retrieved hash chain is invalid: invalid mix digest"
Mar 24 07:41:32 ubp52 sh[24111]: ERROR[03-24|07:41:32.558|eth/handler.go:368]                 Ethereum peer removal failed             peer=818e0548 err="peer not registered"
Mar 24 07:41:49 ubp52 sh[24111]: WARN [03-24|07:41:49.892|eth/downloader/downloader.go:1659]  Invalid header encountered               number=12000031 hash="19eaf2…fb5b97" parent="0ed0e0…436b7a" err="invalid mix digest"
Mar 24 07:41:49 ubp52 sh[24111]: WARN [03-24|07:41:49.892|core/blockchain.go:623]             Rewinding blockchain                     target=12000000
Mar 24 07:41:49 ubp52 sh[24111]: INFO [03-24|07:41:49.893|core/blockchain.go:477]             Loaded most recent local header          number=12000000 hash="4df1f3…2702fb" td=1070480238291937640602 age=2mo1w4h
Mar 24 07:41:49 ubp52 sh[24111]: INFO [03-24|07:41:49.893|core/blockchain.go:478]             Loaded most recent local full block      number=0        hash="d4e567…cb8fa3" td=17179869184            age=51y11mo2w
Mar 24 07:41:49 ubp52 sh[24111]: INFO [03-24|07:41:49.893|core/blockchain.go:479]             Loaded most recent local fast block      number=12000000 hash="4df1f3…2702fb" td=1070480238291937640602 age=2mo1w4h
Mar 24 07:41:49 ubp52 sh[24111]: INFO [03-24|07:41:49.893|core/blockchain.go:481]             Loaded last fast-sync pivot marker       number=12419360
Mar 24 07:41:49 ubp52 sh[24111]: WARN [03-24|07:41:49.893|eth/downloader/downloader.go:1560]  Rolled back chain segment                header=12000000->12000000 fast=12000000->12000000 block=0->0 reason="invalid mix digest"
Mar 24 07:41:49 ubp52 sh[24111]: INFO [03-24|07:41:49.893|eth/downloader/statesync.go:610]    Imported new state entries               count=1   elapsed="1.845µs" processed=153584202 pending=17   trieretry=0 coderetry=0 duplicate=0 unexpected=0
Mar 24 07:41:49 ubp52 sh[24111]: INFO [03-24|07:41:49.893|eth/downloader/queue.go:391]        Downloader queue stats                   receiptTasks=0 blockTasks=0   itemSize=698.79B throttle=8192
Mar 24 07:41:49 ubp52 sh[24111]: WARN [03-24|07:41:49.893|eth/downloader/downloader.go:352]   Synchronisation failed, dropping peer    peer=2a62660cce76343cf2cbb563a6b9032fddae1158a9efd83a5d0a98365e1fac0d err="retrieved hash chain is invalid: invalid mix digest"
Mar 24 07:41:49 ubp52 sh[24111]: ERROR[03-24|07:41:49.893|eth/handler.go:368]                 Ethereum peer removal failed             peer=2a62660c err="peer not registered"
Mar 24 07:41:55 ubp52 sh[24111]: INFO [03-24|07:41:55.862|p2p/dial.go:339]                    Looking for peers                        peercount=2 tried=173 static=0
Mar 24 07:42:06 ubp52 sh[24111]: WARN [03-24|07:42:06.432|eth/downloader/downloader.go:1659]  Invalid header encountered               number=12000033 hash="81494d…f28286" parent="6052fe…0cbbe1" err="invalid mix digest"
Mar 24 07:42:06 ubp52 sh[24111]: WARN [03-24|07:42:06.432|core/blockchain.go:623]             Rewinding blockchain                     target=12000000
Mar 24 07:42:06 ubp52 sh[24111]: INFO [03-24|07:42:06.432|core/blockchain.go:477]             Loaded most recent local header          number=12000000 hash="4df1f3…2702fb" td=1070480238291937640602 age=2mo1w4h
Mar 24 07:42:06 ubp52 sh[24111]: INFO [03-24|07:42:06.432|core/blockchain.go:478]             Loaded most recent local full block      number=0        hash="d4e567…cb8fa3" td=17179869184            age=51y11mo2w
Mar 24 07:42:06 ubp52 sh[24111]: INFO [03-24|07:42:06.432|core/blockchain.go:479]             Loaded most recent local fast block      number=12000000 hash="4df1f3…2702fb" td=1070480238291937640602 age=2mo1w4h
Mar 24 07:42:06 ubp52 sh[24111]: INFO [03-24|07:42:06.432|core/blockchain.go:481]             Loaded last fast-sync pivot marker       number=12419368
Mar 24 07:42:06 ubp52 sh[24111]: WARN [03-24|07:42:06.432|eth/downloader/downloader.go:1560]  Rolled back chain segment                header=12000000->12000000 fast=12000000->12000000 block=0->0 reason="invalid mix digest"
Mar 24 07:42:49 ubp52 sh[24111]: WARN [03-24|07:42:49.766|eth/downloader/downloader.go:352]   Synchronisation failed, dropping peer    peer=2a62660cce76343cf2cbb563a6b9032fddae1158a9efd83a5d0a98365e1fac0d err="retrieved hash chain is invalid: invalid mix digest"
```

**Comment by meowsbits** [comment]
Just some notes about important changes.

- (cmd/utils/flags.go) Legacy datadir migration/path handling for `--testnet` vs `--ropsten` is gone (along with the `--testnet` flag removal)

**Comment by meowsbits** [comment]
Fixme:

- [x] fixup empty `versions=[]`
```
39|eth/backend.go:170]                 Initialising Ethereum protocol           versions=[] network=1 dbversion=8
```

- [x] review and maybe fix genesis block overrides w/r/t berlin, ecbp1100, ecbp1100noDisable, magneto; as below and in `SetupGenesisBlockWithOverride`
```
// ECBP1100NoDisable overrides
	// When this value is *true, ECBP100 will not (ever) be disabled; when *false, it will never be enabled.
	ECBP1100NoDisable *bool `toml:",omitempty"`

	// Berlin block override (TODO: remove after the fork)
	OverrideBerlin *big.Int `toml:",omitempty"`
```

**Comment by meowsbits** [comment]
I cherry-picked https://github.com/etclabscore/core-geth/pull/348/commits/834b4922bb78206fe8aee7650933a0a1256efa73 because the `copydb` command was failing and apparently broken (error was related to apparent invalid protocol version).

**Comment by meowsbits** [comment]
Note for posterity: With the merge commit at https://github.com/etclabscore/core-geth/pull/348/commits/3e3979747d7f1d2687169d0fd95eb1b2ae91c8bd, this branch now includes all changes from 1.10.2, at noted in the changed title of this PR. (Although the branch name is still `...1.10.1...`).

File: core/headerchain.go [hunk 1]
```diff
@@ -129,129 +129,211 @@ func (hc *HeaderChain) GetBlockNumber(hash common.Hash) *uint64 {
...
+		ignored:    len(headers) - len(inserted),
+		imported:   len(inserted),
+		lastHash:   lastHash,
+		lastHeader: lastHeader,
+	}, nil
+}
 
 func (hc *HeaderChain) ValidateHeaderChain(chain []*types.Header, checkFreq int) (int, error) {
 	// Do a sanity check that the provided chain is actually ordered and linked
+	// NOTE(meowsbits): These checks for nil heads are core-geth specific.
```
Comment (ziogaschr) [inline core/headerchain.go:301]: @meowsbits is this something you are willing to check?

**Comment by ziogaschr** [comment]
@meowsbits & @iquidus I reviewed all the file changes (not tests) and it looks that we haven't missed anything on merge. It LGTM for an RC release.

File: core/headerchain.go (hunk 1 above)
Comment (meowsbits) [reply core/headerchain.go:301]: Ah, yeap -- I did check that; all good. 

- https://github.com/etclabscore/core-geth/blob/merge/foundation-release%2F1.10.1-resolved/core/headerchain.go#L305-L311
- https://github.com/etclabscore/core-geth/blob/merge/foundation-release/1.10.1-resolved/light/txpool_test.go#L137

Comment (meowsbits) [reply core/headerchain.go:301]: See https://github.com/etclabscore/core-geth/commit/5c2f6c25c4410b267add0038307da120afe89259

**Comment by meowsbits** [comment]
The Travis CI test failures look to me to be typically spurious (`A7-looksgoodtestsfail` :joy:). I'll run a `devp2p discv4 crawl` right now to make sure we won't break the downstream DNS updater, then AOK to me too.

Testing at f912896.

- [x] _Update:_ Looks OK. :heavy_check_mark:

**Comment by meowsbits** [comment]
Closing in favor of #376 _Release/1.12 RC-1_.

**Pull request closed without merging**